near-rust-allocator-proxy/src/allocator.rs

```diff
@@ -237,16 +237,19 @@ unsafe impl<A: GlobalAlloc> GlobalAlloc for ProxyAllocator<A> {
             }
             in_trace.set(0);
         });
+
+        let (new_layout, new_size) = layout.extend(Layout::new::<AllocHeader>()).unwrap();
```

Comment:
Presume we call the type for which `layout` was constructed a `Original`. This specific layout will allocate a memory region that contains `Original` at offset 0 and `AllocHeader` at offset `new_size`, that is:

```
+ ---------------------- + ------- + ----------- +
| Original               | padding | AllocHeader |
+ ---------------------- + ------- + ----------- +
| 0..size_of::<Original> |         | new_size..  |
+ ---------------------- + ------- + ----------- +
```

Right now this code is writing out `AllocHeader` to offset 0, so the appropriate way to construct this layout would be like this:

```rust
let (new_layout, offset) = Layout::new::<AllocHeader>().extend(layout);
```

Once that's done, the pointer to return should be calculatable the following way:

```rust
res.add(offset)
```

instead of what is being done currently.

Similarly for deallocation (`ptr.sub(offset)` would give back the original pointer.)

Reply:
> Presume we call the type for which `layout` was constructed a `Original`. This specific layout will allocate a memory region that contains `Original` at offset 0 and `AllocHeader` at offset `new_size`, that is:
> 
> ```
> + ---------------------- + ------- + ----------- +
> | Original               | padding | AllocHeader |
> + ---------------------- + ------- + ----------- +
> | 0..size_of::<Original> |         | new_size..  |
> + ---------------------- + ------- + ----------- +
> ```
> 
> Right now this code is writing out `AllocHeader` to offset 0, so the appropriate way to construct this layout would be like this:
> 
> ```rust
> let (new_layout, offset) = Layout::new::<AllocHeader>().extend(layout);
> ```
> 
> Once that's done, the pointer to return should be calculatable the following way:
> 
> ```rust
> res.add(offset)
> ```
> 
> instead of what is being done currently.
> 
> Similarly for deallocation (`ptr.sub(offset)` would give back the original pointer.)

I understand, that approach will work too. 

However, if someone allocates a large vector over a page long. Linux by default will not allocate all the data.
Let's say `Vec` allocates a region of memory, but only uses front of that region. If we put the header at the end, in the worst case, Linux may need to allocate 2 pages instead of 1, causing higher memory usage.

Reply:
I think we're talking past each other, for some reason.

What I'm saying is that the order of arguments to `Layout::extend` matter. As written currently it produces a layout for an allocation where `AllocHeader` is at the end of the original allocation and that you want to reverse the order of the arguments so that the allocation layout and where you want to place data actually match.

Reply:
In other words:

```rust
let (new_layout, offset) = Layout::new::<AllocHeader>().extend(layout);
// produces...
// + ----------- + ------- + ---------------------- +
// | AllocHeader | padding | Original               |
// + ----------- + ------- + ---------------------- +
// where `Original` starts at `offset`
```

and

```rust
let (new_layout, offset) = layout.extend(Layout::new::<AllocHeader>()).unwrap(); 
// produces
// + ---------------------- + ------- + ----------- +
// | Original               | padding | AllocHeader |
// + ---------------------- + ------- + ----------- +
// where `AllocHeader` starts at `offset`
```

Hopes that explains what I'm trying to say.

Reply:
@nagisa Oh, I see that makes sense. I agree, reversing the order of arguments to `Layout::extend` will make the code cleaner.

```rust
Layout::new::<AllocHeader>().extend(layout).unwrap();
```
rather than
```
layout.extend(Layout::new::<AllocHeader>()).unwrap(); 
```

Reply:
@nagisa Actually, now that I think about it. The previous code was wrong.

If someone requested to allocate `(1024 - 32)` bytes with 64 bytes alignment, and assuming the header is 32 bytes.

Then `layout.extend(Layout::new::<AllocHeader>()).unwrap()` would allocate 1024, and the returned pointer wouldn't be aligned to 64 bytes.

`Layout::new::<AllocHeader>().extend(layout).unwrap();` would allocate (64 + 1024), making the alignment correct.